File: slp/backend-dotnet/Features/Session/SessionRepository.cs
using backend_dotnet.Data;
using Microsoft.EntityFrameworkCore;

namespace backend_dotnet.Features.Session;

public class SessionRepository : ISessionRepository
{
    private readonly AppDbContext _db;

    public SessionRepository(AppDbContext db)
    {
        _db = db;
    }

    public async Task CreateAsync(Session session)
    {
        _db.Sessions.Add(session);
        await _db.SaveChangesAsync();
    }

    public async Task<Session?> GetByTokenHashAsync(string hash)
    {
        return await _db.Sessions.FirstOrDefaultAsync(x => x.TokenHash == hash);
    }

    public async Task RevokeAsync(string sessionId)
    {
        var session = await _db.Sessions.FindAsync(sessionId);

        if (session == null)
            return;

        session.Revoked = true;

        await _db.SaveChangesAsync();
    }

    public async Task RevokeAllForUserAsync(int userId)
    {
        var sessions = await _db.Sessions
            .Where(s => s.UserId == userId && !s.Revoked)
            .ToListAsync();
        foreach (var session in sessions)
        {
            session.Revoked = true;
        }
        await _db.SaveChangesAsync();
    }
}

File: slp/backend-java/src/main/java/com/app/backendjava/features/session/SessionRepository.java
package com.app.backendjava.features.session;

import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.data.jpa.repository.Modifying;
import org.springframework.data.jpa.repository.Query;
import org.springframework.data.repository.query.Param;
import org.springframework.stereotype.Repository;
import org.springframework.transaction.annotation.Transactional;

import java.util.Optional;

@Repository
public interface SessionRepository extends JpaRepository<Session, String> {

    Optional<Session> findByTokenHash(String tokenHash);

    @Modifying
    @Transactional
    @Query("UPDATE Session s SET s.revoked = true WHERE s.id = :sessionId")
    void revokeAsync(@Param("sessionId") String sessionId);

    @Modifying
    @Transactional
    @Query("UPDATE Session s SET s.revoked = true WHERE s.userId = :userId AND s.revoked = false")
    void revokeAllForUserAsync(@Param("userId") Integer userId);

    @Modifying
    @Transactional
    @Query("UPDATE Session s SET s.revoked = true WHERE s.userId = :userId AND s.revoked = false AND s.id != :exceptSessionId")
    void revokeAllForUserExceptAsync(@Param("userId") Integer userId, @Param("exceptSessionId") String exceptSessionId);
}